README.md
# VitalSigns [![Build Status](https://travis-ci.org/TomFrost/node-vitalsigns.svg?branch=master)](https://travis-ci.org/TomFrost/node-vitalsigns)
Powerful and customizable application health monitoring

Sponsored by [Leadnomics](http://www.leadnomics.com).

## Installation
In your project folder, type:

	npm install vitalsigns

## Basic Usage
Load up VitalSigns and include a few of the built-in monitors:

	var VitalSigns = require('vitalsigns'),
		vitals = new VitalSigns();

	vitals.monitor('cpu');
	vitals.monitor('mem', {units: 'MB'});
	vitals.monitor('tick');

We know the CPU hitting 100% is bad news. So is the tick.maxMs clearing 500.

	vitals.unhealthyWhen('cpu', 'usage').equals(100);
	vitals.unhealthyWhen('tick', 'maxMs').greaterThan(500);

Let's get application-specific.  We'll monitor connections to a game and go
unhealthy when they cross 200.

	vitals.monitor({
		connections: function() { return Game.getConnections(); }
	}, {name: 'game'});

	vitals.unhealthyWhen('game', 'connections').greaterThan(200);

We need to know when we go unhealthy...

	vitals.on('healthChange', function(healthy, failedChecks) {
		console.log("Server is " + (healthy ? 'healthy' : 'unhealthy') +
			".  Failed checks:", failedChecks);
	});

And we have a load balancer hitting /health and looking for a non-200
response from express.

	app.get('/health', vitals.express);


## The API
### Constructor
VitalSigns must be instantiated to use, and can optionally receive a set of
options:

	var vitals = new VitalSigns({
		autoCheck: 5000,
		httpHealthy: 200,
		httpUnhealthy: 503
	});

#### autoCheck
*Default: false.* The number of milliseconds to wait between automatic health
checks, or boolean false to disable.  Alternatively, 'true' can be specified to
auto-check every 5 seconds.

#### httpHealthy
*Default: 200.* The HTTP response code to send back in the VitalSigns.express
HTTP endpoint when the server is healthy.

#### httpUnhealthy
*Default: 503.* The HTTP response code to send back in the VitalSigns.express
HTTP endpoint when the server is unhealthy.

### Monitors
Without stats to monitor, VitalSigns does nothing!  In order for it to be
useful, VitalSigns must be told what monitors to use.  These can be the name
of a built in monitor (cpu, mem, tick, and uptime), the name of a library in
your project's node_modules folder, an object, or a function.  See the "Kinds
of Monitors" section below for more details.

Monitors are registered with:

**instance.monitor({string|object|function} monitor, [{object} options])**

Options are optional.  If specified, they should be a Javascript object of
key/value pairs specific to the monitor being loaded.  In addition, a *name*
field can be specified to override the default name for the module.  For
example, the following will cause CPU reports to be grouped under 'foo' instead
of 'cpu':

	instance.monitor('cpu', {name: 'foo'});

### Health constraints
Before or after monitors are loaded, VitalSigns can be told what values are
considered unhealthy.  The syntax starts with:

	instance.unhealthyWhen(<monitor>, <field>)

and the following can be added to the end to complete the definition:

	.equals(<value>)
	.greaterThan(<value>)
	.lessThan(<value>)
	.not // chainable with one of the above

For example, to mark the instance as unhealthy if monitor "foo" has a field
named "bar" that's equal to or less than 5, use:

	instance.unhealthyWhen('foo', 'bar').not.greaterThan(5);

Many constraints can be defined on one instance of VitalSigns.

### Function calls
Besides the above, the following calls are available on VitalSigns instances:

#### destroy()
Destructs the instance, terminating autoChecks and any intervals set by any
of the attached monitors.  Also removes all event listeners.  This is handy
to do before shutdown to eliminate any ongoing process that might prevent
the process from exiting.

#### express
A function to be passed to Express as the endpoint for a route.  This function
will return HTTP 200 or 503 by default to represent healthy and unhealthy,
respectively.  It will also provide the full health report returned by
`getReport()` as a JSON string with *Content-type: application/json*.  Example:

	app.get('/ping', instance.express);

#### hapi
A function to be passed to Hapi as the handler for a route.  This function
will return HTTP 200 or 503 by default to represent healthy and unhealthy,
respectively.  It will also provide the full health report returned by
`getReport()` as a JSON string with *Content-type: application/json*.  Example:

	server.route({method: 'GET', path: '/ping', handler: instance.hapi});

#### {array} getFailed()
Returns an array of strings describing the constraints that failed the last
time `isHealthy()` was called.  Array will be empty if the instance was
healthy as of the last check.

#### {object} getReport(options)
Returns a Javascript object with the health report: each monitor name as keys,
with the value being another Javascript object mapping each monitor's fields to
their values.  This function also attaches a 'healthy' field at the root level
with a boolean true or false representing whether the instance is healthy based
on this report.

Sample report:

	{
		cpu: {
			usage: 50,
			loadAvg1: 0.09,
			loadAvg5: 0.80,
			loadAvg15: 1.29
		},
		healthy: true
	}

##### {boolean} options.flatten (default: false)
Set to true to flatten the report object to a single level of keys by
concatenating nested key names.  Example:

	{
		"cpu.usage": 50,
		"cpu.loadAvg1": 0.09,
		"cpu.loadAvg5": 0.80,
		"cpu.loadAvg15": 1.29,
		"healthy": true
	}

##### {string} options.separator (default: ".")
If flatten is true, this is the string used to separate joined key names. 

#### {boolean} isHealthy()
Returns true or false based on whether the instance is healthy.  Also fires up
to two events; see "Events" section below.

### Events
VitalSigns fires up to two events when the health is checked:

#### healthCheck (healthy, report, failures)
Fires every time `isHealthy()` is called, which includes autoChecks and calls
to `getReport()`.

- **healthy** *boolean:* true if healthy; false if not
- **report** *object:* The raw monitor reports in a Javascript object, grouped
by monitor name
- **fails** *array:* An array of strings describing the individual constraints
that failed and caused the unhealthy status.  Array has length=0 if healthy.

#### healthChange (healthy, report, failures)
Fires when a health check is performed that causes the instance to switch from
healthy to unhealthy, or vice versa.  Provides the same arguments as the
healthCheck event.

## Kinds of Monitors
VitalSigns-compatible monitors come in all shapes and sizes.  For distributed
monitors, an object with 'name' and 'report' fields is highly recommended. Here
are a few of the configurations:

### Object with static fields

	module.exports = {
		appName: "My Awesome App",
		hostname: os.hostname()
	};

### Object with dynamic fields

	module.exports = {
		connections: function() {
			return myApp.getConnections();
		}
	};

### Named monitor with report function

	module.exports = {
		name: 'MyMonitor',
		report: function() {
			return {
				connections: myApp.getConnections();
				lastConnection: myApp.getLastConnectionTime();
			};
		}
	};

### Function returning one of the above objects

	module.exports = function(options) {
		if (!options)
			options = {};
		return {
            name: 'MyMonitor',
            report: function() {
                return {
                    connections: myApp.getConnections();
                    lastConnection: myApp.getLastConnectionTime(options.dateFormat);
                };
            }
        };
	}

## Built-in monitors
VitalSigns comes with a small number of application-unspecific monitors to
report on general server and process health.  They are:

### cpu
Monitors CPU usage and load.  Provides fields:

- **usage:** Percent of CPU being used at this moment
- **loadAvg1:** The 1-minute load average
- **loadAvg5:** The 5-minute load average
- **loadAvg15:** The 15-minute load average

Options:

- **sampleTime** *default 1000:* The milliseconds over which to sample CPU
usage
- **updateTime** *default 5000:* The milliseconds to wait between samples

### mem
Monitors memory usage.  Provides fields:

- **free:** Amount of memory currently available
- **process:** Amount of memory currently being used by this node process.

Options:

- **units** *default B:* The units in which to display RAM sizes.  Legal
options are B, KB, MB, GB, TB, and PB.  If you can measure your RAM in PB,
I accept RAM donations.

### tick
Monitors the speed of the event loop.  For a Node.js app, this is the single
most important set of statistics available.  Provides fields:

- **avgMs:** Average number of milliseconds required to loop through the
specified number of ticks
- **maxMs:** For the last batch of samples, the milliseconds required to loop
through the slowest batch of ticks
- **perSec:** An estimate of how many individual ticks are being completed per
second.

Options:

- **window** *default 10000:* The number of milliseconds for which to collect
batches of tick measurements before they are averaged.
- **batch** *default 1000:* The number of ticks to be timed in a single batch.
Batches of this size are what get processed to create `avgMs` and `maxMs`.
- **freq** *default 50:* The number of milliseconds to pause between collecting
batches of ticks.

### uptime
Monitors application uptime.  Provides fields:

- **sys:** The number of seconds for which the server has been online
- **proc:** The number of seconds for which this node process has been running

## License
VitalSigns is distributed under the MIT license.

## Credits
VitalSigns was created by Tom Frost at Leadnomics in 2013.


CHANGELOG.md
# VitalSigns
Powerful and customizable application health monitoring

##ChangeLog

### v0.4.3
- Fixed: Support the official Express 4 API

### v0.4.2
- Added: Travis CI integration
- Fixed: Express endpoint now sends (statusCode, body) instead of
(body, statusCode). This avoids a warning in Express 4.x, and is
backward-compatible. 

### v0.4.1
- Fixed: Calling getReport() no longer generates a second report to run the
health check
- Fixed: Various documentation improvements, especially for IDEs that enforce
JSDoc declared types

### v0.4.0
- Added: JSDoc on all functions in the core
- Added: Ability to get flattened reports

### v0.3.0
- Added: Hapi route handler, alongside the existing express option. Credits
to [mrlannigan](https://github.com/mrlannigan).

### v0.2.2
- Fixed: Reports overwrite function values in monitors

### v0.2.1
- Fixed: Eliminate missing repository warning from NPM

### v0.2.0
- README clarifications
- Feature: Allow 'report' to link to a static value so that non-namespaced
keys can be attached to a report
- Feature: Allow constraints to be built around non-namespaced keys
- Fixed: Error in express endpoint when accessing the unhealthy HTTP code.
- Fixed: uptime monitor was not in the index

### v0.1.0
- **Initial Release**


lib/VitalSigns.js
/*
 * VitalSigns
 * Copyright 2013 Tom Frost
 */

const DEFAULT_AUTOCHECK = 5000;
const DEFAULT_HTTP_HEALTHY = 200;
const DEFAULT_HTTP_UNHEALTHY = 503;

var builtIn = require('./monitors'),
	ConstraintWrapper = require('./ConstraintWrapper'),
	Comparators = require('./Comparators'),
	util = require('util'),
	events = require('events'),
	objUtil = require('./util/Object');

/**
 * Constructs a new instance of VitalSigns
 * @param {Object} options An optional mapping of options to values
 * @param {boolean|number} [options.autoCheck=false] The number of milliseconds
 *      to wait between automatic health checks, or false to disable.
 *      Alternatively, true can be specified to auto-check every 5 seconds.
 * @param {number} [options.httpHealthy=200] The HTTP response code to send
 *      back in the HTTP endpoints when the server is healthy
 * @param {number} [options.httpUnhealthy=503] The HTTP response code to send
 *      back in the HTTP endpoints when the server is unhealthy
 * @constructor
 */
var VitalSigns = function(options) {
	if (!options)
		options = {};
	this._opts = options;
	this._monitors = {};
	this._constraints = [];
	this._failed = [];
	this._autoCheck = null;
	this._lastCheckHealthy = true;
	if (options.autoCheck) {
		this._autoCheck = setInterval(this.isHealthy.bind(this),
			options.autoCheck === true ? DEFAULT_AUTOCHECK :
				options.autoCheck);
	}

	this.express = expressEndpoint.bind(this);
	this.hapi = hapiEndpoint.bind(this);
};
util.inherits(VitalSigns, events.EventEmitter);

/**
 * Pushes a health constraint onto this instance's constraint array. Health
 * constraints define scenarios in which VitalSigns will consider the
 * application to be in an unhealthy state.
 * @param {{monitor: string, field: string, comparator: string, value: *,
 *      [negate]: boolean}} constraint A constraint object
 * @param {string} constraint.monitor The name of the monitor containing the
 *      field to be constrained
 * @param {string} constraint.field The name of the field to be constrained
 * @param {string} constraint.comparator The comparator to use when comparing
 *      the field's value with the constraint value.  Valid comparators are:
 *      'equal', 'greater', and 'less'.
 * @param {*} constraint.value The value against which the field should be
 *      compared
 * @param {boolean} [constraint.negate=false] true to negate the outcome of
 *      the comparison; false or omitted to use the comparison result
 */
VitalSigns.prototype.addConstraint = function(constraint) {
	this._constraints.push(constraint);
};

/**
 * Destroys this VitalSigns instance. Instances can not be re-activated after
 * being destroyed.  This will halt the 'autoCheck' timer (if specified in the
 * constructor options), remove all event listeners, and call the destroy
 * method for any added monitors.
 */
VitalSigns.prototype.destroy = function() {
	if (this._autoCheck) {
		clearInterval(this._autoCheck);
		this._autoCheck = null;
	}
	this.removeAllListeners();
	for (var modName in this._monitors) {
		if (this._monitors.hasOwnProperty(modName) &&
				this._monitors[modName].destroy)
			this._monitors[modName].destroy();
	}
};

/**
 * Retrieves an array of human-readable messages that define the specific
 * health constraints that failed when running the last health check.
 * @returns {Array<string>} An array of failure messages
 */
VitalSigns.prototype.getFailed = function() {
	return this._failed;
};

/**
 * Gets a report of all monitors, their fields, and the values of those fields,
 * compiled into Javascript object form.  Additionally, a 'healthy' field is
 * attached. This field will be boolean true if all health constraints passed;
 * false otherwise.
 * @param {Object} [opts] A mapping of options to customize this report
 * @param {boolean} [opts.flatten=false] true to flatten the report object down
 *      to a single level by concatenating nested key names; false to keep the
 *      default hierarchical format.
 * @param {string} [opts.separator="."] If flatten is true, this string will be
 *      used to separate key names when they are concatenated together.
 * @returns {{}} A full health report
 */
VitalSigns.prototype.getReport = function(opts) {
	if (!opts) opts = {};
	var report = this._getRawReport();
	report.healthy = this.isHealthy(report);
	if (opts.flatten)
		report = objUtil.flatten(report, null, opts.separator);
	return report;
};

/**
 * Generates a health report and checks each health constraint against it.
 * Any constraints that fail will be added to the 'failed' array in the form of
 * a human-readable failure message, which can be retrieved with
 * {@link #getFailed}.
 * @param {{}} [report] A report object on which to run the health constraints.
 *      If omitted, this function will generate a health report automatically.
 * @returns {boolean} true if all health constraints passed; false otherwise.
 */
VitalSigns.prototype.isHealthy = function(report) {
	if (!report)
		report = this._getRawReport();
	var fails = [],
		failMsg = '%s.%s %s %s',
		_this = this;
	this._constraints.forEach(function(constraint) {
		if (_this._constraintHit(constraint, report)) {
			var msg = util.format(failMsg, constraint.monitor, constraint.field,
				Comparators[constraint.comparator].text,
				constraint.value + '');
			fails.push(msg);
		}
	});
	this._failed = fails;
	var healthy = !fails.length;
	this.emit('healthCheck', healthy, report, fails);
	if (healthy != this._lastCheckHealthy) {
		this.emit('healthChange', healthy, report, fails);
		this._lastCheckHealthy = healthy;
	}
	return healthy;
};

/**
 * Defines a new health constraint in a chainable, more easily readable format.
 * When called with the monitor name and field name of concern, a wrapper is
 * returned that allows the constraint to be built out with function calls.
 * For example:
 *
 *      vitals.unhealthyWhen('cpu', 'usage').greaterThan(90);
 *
 * The constraint can be negated by adding 'not' to the chain as well:
 *
 *      vitals.unhealthyWhen('app', 'uncaughtErrors').not.equals(0);
 *
 * Valid comparison functions are greaterThan(number), lessThan(number), and
 * equals(*).
 * @param {string} monitorName The name of the monitor containing the field to
 *      be checked
 * @param {string} fieldName The name of the field to be checked in this health
 *      constraint
 * @returns {ConstraintWrapper} A wrapper object containing chainable functions
 *      that define the constraint to be added.
 */
VitalSigns.prototype.unhealthyWhen = function(monitorName, fieldName) {
	if (!this._monitors.hasOwnProperty(monitorName))
		throw new Error('Monitor "' + monitorName + '" is not loaded.');
	return new ConstraintWrapper(this, monitorName, fieldName);
};

/**
 * Instructs VitalSigns to monitor some value or set of values.  Monitors can
 * take on various forms:
 *
 *      Built-in: VitalSigns contains a set of built-in monitors (see the
 *          'monitors' folder) which can be added by name.  For example,
 *          vitals.monitor('cpu').  Additionally, any health monitor that can
 *          be require()d by name can also be passed.  For example, 'appHealth'
 *          would trigger vitalSigns to call this.monitor(require('appHealth'))
 *
 *      Object with report function: An object containing a function called
 *          'report' that, when run, returns an object containing the fields
 *          and values to be added to the health report.  Optionally, a default
 *          monitor name can be specified in the 'name' field.  For example:
 *
 *              {
 *                  name: 'app',
 *                  report: function() {
 *                      return {
 *                          connections: myApp.connections.length,
 *                          shuttingDown: myApp.isShuttingDown()
 *                      };
 *                  }
 *              }
 *
 *          The 'report' function will be called every time a health report
 *          needs to be generated.
 *
 *      Object with included fields: A object containing all the fields to be
 *          monitored.  Fields can be mapped to either static values, or
 *          functions that will be called to generate that field's value every
 *          time a report is run.  For example:
 *
 *              {
 *                  connections: function() { return myApp.connections.length },
 *                  host: os.hostname()
 *              }
 *
 *      Generated: A function that returns one of the monitor object types
 *          defined above. The provided 'options' object will be passed to it
 *          when it's called. This function will be called only once.
 * @param {string|{[name]: string}|function} monitor A string, object, or
 *      function as defined above
 * @param {Object} options An options mapping to be considered when adding the
 *      given monitor.  If the given monitor is a function, this object will
 *      be passed to it as its first argument when the function is called.
 * @param {string} [options.name=default] The name under which to store this
 *      monitor.  If not specified, module.name will be used (if present).  If
 *      that does not exist, 'default' will be used.
 */
VitalSigns.prototype.monitor = function(monitor, options) {
	if (!options)
		options = {};
	if (typeof monitor == 'string')
		monitor = this._getModuleByName(monitor);
	if (typeof monitor == 'function')
		monitor = monitor(options);
	if (typeof monitor != 'object')
		throw new Error("Monitor must be either a function or object.");
	var name = options.name || monitor.name || 'default';
	if (!this._monitors[name])
		this._monitors[name] = {};
	for (var i in monitor) {
		if (monitor.hasOwnProperty(i))
			this._monitors[name][i] = monitor[i];
	}
};

/**
 * Determines whether the provided health report violates a set health
 * constraint.
 * @param {{monitor: string, field: string, comparator: string, value: *,
 *      [negate]: boolean}} constraint The constraint object to check against
 *      the provided report
 * @param {Object} report A health report
 * @returns {boolean} true if the constraint is violated; false otherwise
 * @private
 */
VitalSigns.prototype._constraintHit = function(constraint, report) {
	var value,
		comparator,
		result;

	if (!report[constraint.monitor])
		return false;
	if (!report[constraint.monitor].hasOwnProperty(constraint.field)) {
		if (typeof report[constraint.monitor] === 'object')
			return false;
		else
			value = report[constraint.monitor];
	}
	else
		value = report[constraint.monitor][constraint.field];

	comparator = Comparators[constraint.comparator].func;
	result = comparator(value, constraint.value);

	return constraint.negate ? !result : result;
};

/**
 * Loads a monitor by first checking to see whether the name matches a built-in
 * module, or, if not, attempting to require() that monitor name.
 * @param {string} name The name of the module to be loaded
 * @returns {Object} The loaded monitor
 * @private
 */
VitalSigns.prototype._getModuleByName = function(name) {
	var mod;
	if (builtIn.hasOwnProperty(name))
		mod = builtIn[name];
	else
		mod = require(name);
	if (!mod.name)
		mod.name = name;
	return mod;
};

/**
 * Generates the report for an individual health monitor.
 * @param {{[report]: function}|function} monitor The health monitor from which
 *      the report should be generated
 * @returns {Object} The generated report
 * @private
 */
VitalSigns.prototype._getModuleReport = function(monitor) {
	var report;
	if (typeof monitor == 'function')
		return this._getModuleReport(monitor());
	if (monitor.report && typeof monitor.report == 'function')
		report = monitor.report();
	else
		report = shallowClone(monitor);
	for (var field in report) {
		if (report.hasOwnProperty(field)) {
			if (typeof report[field] == 'function')
				report[field] = report[field]();
		}
	}
	return report;
};

/**
 * Generates a health report by getting the report of each individual monitor
 * and adding each to a parent object, using their monitor name as the key
 * name.  Note that this function does not invoke a health check; it simply
 * generates a report.
 * @returns {{}} The raw health report
 * @private
 */
VitalSigns.prototype._getRawReport = function() {
	var report = {};
	for (var modName in this._monitors) {
		if (this._monitors.hasOwnProperty(modName))
			report[modName] = this._getModuleReport(this._monitors[modName]);
	}
	return report;
};

/**
 * An endpoint suitable to attach to an Express route.  Note that this function
 * must be bound to a VitalSigns instance before being called.
 * @param req
 * @param res
 */
function expressEndpoint(req, res) {
	var report = this.getReport(),
		status = report.healthy ?
			(this._opts.httpHealthy || DEFAULT_HTTP_HEALTHY) :
			(this._opts.httpUnhealthy || DEFAULT_HTTP_UNHEALTHY);
	if (res.status)
		res.status(status).json(report);
	else
		res.json(status, report);
}

/**
 * An endpoint suitable to attach to a Hapi route.  Note that this function
 * must be bound to a VitalSigns instance before being called.
 * @param req
 * @param reply
 */
function hapiEndpoint(req, reply) {
	var report = this.getReport();
	reply(report).code(report.healthy ?
		(this._opts.httpHealthy || DEFAULT_HTTP_HEALTHY) :
		(this._opts.httpUnhealthy || DEFAULT_HTTP_UNHEALTHY));
}

/**
 * Generates a shallow, top-level clone of the given object.  Note that only
 * javascript natives will be cloned; for all others, the object reference will
 * be copied.
 * @param {Object} obj The object to be shallow-cloned
 * @returns {Object} The cloned object
 */
function shallowClone(obj) {
	var cloned = {};
	if (typeof obj == 'object') {
		for (var key in obj) {
			if (obj.hasOwnProperty(key))
				cloned[key] = obj[key];
		}
		return cloned;
	}
	return obj;
}

module.exports = VitalSigns;
